File: backend/opencode/messages.go
package opencode

import (
	"encoding/json"
	"fmt"
	"strings"
)

const toolIncompleteMsg = "Error: tool call was not completed"

// StripResponseFields removes response-only metadata before sending messages to
// the provider. OpenCode Go rejects requests that include usage on messages.
func StripResponseFields(msgs []Message) []Message {
	if len(msgs) == 0 {
		return msgs
	}
	out := make([]Message, len(msgs))
	for i, msg := range msgs {
		msg.Usage = nil
		out[i] = msg
	}
	return out
}

// PrepareRequestMessages sanitizes session history for chat/completions requests.
func PrepareRequestMessages(msgs []Message, model string) []Message {
	return stripImagesIfNeeded(
		RepairToolMessages(
			SanitizeToolCallArguments(NormalizeMessages(StripResponseFields(msgs), model)),
		),
		model,
	)
}

// SanitizeToolCallArguments ensures assistant tool_call arguments are valid JSON
// objects. Providers reject replayed history when the model emitted malformed
// JSON (e.g. raw newlines inside string literals). The tool result already
// captured the parse error locally, so replacing bad arguments with {} is safe.
func SanitizeToolCallArguments(msgs []Message) []Message {
	if len(msgs) == 0 {
		return msgs
	}
	out := make([]Message, len(msgs))
	for i, msg := range msgs {
		if msg.Role != "assistant" || len(msg.ToolCalls) == 0 {
			out[i] = msg
			continue
		}
		copied := msg
		copied.ToolCalls = make([]ToolCall, len(msg.ToolCalls))
		for j, tc := range msg.ToolCalls {
			copied.ToolCalls[j] = tc
			copied.ToolCalls[j].Function.Arguments = validToolArgumentsJSON(tc.Function.Arguments)
		}
		out[i] = copied
	}
	return out
}

func validToolArgumentsJSON(raw string) string {
	raw = strings.TrimSpace(raw)
	if raw == "" {
		return "{}"
	}
	var obj map[string]json.RawMessage
	if err := json.Unmarshal([]byte(raw), &obj); err != nil || obj == nil {
		return "{}"
	}
	normalized, err := json.Marshal(obj)
	if err != nil {
		return "{}"
	}
	return string(normalized)
}

func stripImagesIfNeeded(msgs []Message, model string) []Message {
	if SupportsImages(model) {
		return msgs
	}
	out := make([]Message, len(msgs))
	for i, msg := range msgs {
		blocks := ContentBlocks(msg)
		if len(blocks) == 0 {
			out[i] = msg
			continue
		}

		var filtered []ContentBlock
		var imageCount int
		for _, b := range blocks {
			if isImageContentBlock(b) {
				imageCount++
			} else {
				filtered = append(filtered, b)
			}
		}
		if imageCount == 0 {
			out[i] = msg
			continue
		}

		note := fmt.Sprintf("[%d image(s) omitted — current model does not support images.]", imageCount)
		if len(filtered) > 0 && filtered[0].Type == "text" {
			filtered[0].Text = filtered[0].Text + "\n" + note
		} else {
			filtered = append([]ContentBlock{{
				Type: "text",
				Text: note,
			}}, filtered...)
		}

		if len(filtered) == 0 {
			msg.Content = nil
		} else if len(filtered) == 1 && filtered[0].Type == "text" {
			msg.Content = StringContent(filtered[0].Text)
		} else {
			msg.Content = BlocksContent(filtered)
		}
		out[i] = msg
	}
	return out
}

func isImageContentBlock(b ContentBlock) bool {
	return b.Type == "image_url" || b.Type == "image"
}

// RepairToolMessages ensures every assistant tool_call has a matching tool response.
// OpenAI-compatible APIs reject requests when tool_calls are missing replies.
func RepairToolMessages(msgs []Message) []Message {
	out := make([]Message, 0, len(msgs))

	for i := 0; i < len(msgs); i++ {
		msg := msgs[i]
		if msg.Role == "assistant" && len(msg.ToolCalls) > 0 {
			for j := range msg.ToolCalls {
				if msg.ToolCalls[j].ID == "" {
					msg.ToolCalls[j].ID = fmt.Sprintf("call_%d_%d", i, j)
				}
			}
		}

		out = append(out, msg)

		if msg.Role != "assistant" || len(msg.ToolCalls) == 0 {
			continue
		}

		required := make([]ToolCall, len(msg.ToolCalls))
		copy(required, msg.ToolCalls)

		answered := make(map[string]bool)
		i++
		for i < len(msgs) && msgs[i].Role == "tool" {
			tm := msgs[i]
			if tm.ToolCallID == "" {
				tm.ToolCallID = required[0].ID
			}
			if !answered[tm.ToolCallID] {
				out = append(out, tm)
				answered[tm.ToolCallID] = true
			}
			i++
		}
		i--

		for _, tc := range required {
			if answered[tc.ID] {
				continue
			}
			out = append(out, Message{
				Role:       "tool",
				ToolCallID: tc.ID,
				Name:       tc.Function.Name,
				Content:    StringContent(toolIncompleteMsg),
			})
		}
	}

	return out
}

File: backend/opencode/messages.ts
import { blocks_content, content_blocks, string_content, content_string, type content_block, type message, type tool_call } from "./types";
import { normalize_messages, is_minimax_model } from "./thinking";
import { supports_images } from "./models";

export const CompactionSummaryPrefix =
  "The conversation history before this point was compacted into the following summary:\n\n<summary>\n";
export const CompactionSummarySuffix = "\n</summary>";
export const BranchSummaryPrefix =
  "The following is a summary of a branch that this conversation came back from:\n\n<summary>\n";
export const BranchSummarySuffix = "\n</summary>";

export const bash_execution_to_text = (msg: any): string => {
  let text = `Ran \`${msg.command || ""}\`\n`;
  if (msg.output) {
    text += `\`\`\`\n${msg.output}\n\`\`\``;
  } else {
    text += "(no output)";
  }
  if (msg.cancelled) {
    text += "\n\n(command cancelled)";
  } else if (msg.exitCode !== null && msg.exitCode !== undefined && msg.exitCode !== 0) {
    text += `\n\nCommand exited with code ${msg.exitCode}`;
  }
  if (msg.truncated && msg.fullOutputPath) {
    text += `\n\n[Output truncated. Full output: ${msg.fullOutputPath}]`;
  }
  return text;
};

export const convert_to_llm = (messages: message[]): message[] => {
  const out: message[] = [];
  for (const m of messages) {
    switch (m.role) {
      case "bashExecution": {
        const anyMsg = m as any;
        if (anyMsg.excludeFromContext) {
          break;
        }
        out.push({
          role: "user",
          content: string_content(bash_execution_to_text(m)),
        });
        break;
      }
      case "custom": {
        out.push({
          role: "user",
          content: m.content,
        });
        break;
      }
      case "compactionSummary": {
        const content = CompactionSummaryPrefix + content_string(m) + CompactionSummarySuffix;
        out.push({
          role: "user",
          content: string_content(content),
        });
        break;
      }
      case "branchSummary": {
        const content = BranchSummaryPrefix + content_string(m) + BranchSummarySuffix;
        out.push({
          role: "user",
          content: string_content(content),
        });
        break;
      }
      default:
        out.push(m);
        break;
    }
  }
  return out;
};

export const tool_incomplete_msg = "Error: tool call was not completed";

export const strip_response_fields = (msgs: message[]): message[] => {
  if (msgs.length === 0) return msgs;
  return msgs.map((msg) => ({ ...msg, usage: undefined }));
};

export const prepare_request_messages = (msgs: message[], model: string): message[] =>
  strip_images_if_needed(
    repair_tool_messages(
      sanitize_tool_call_arguments(
        strip_minimax_reasoning_details(normalize_messages(strip_response_fields(convert_to_llm(msgs)), model), model),
      ),
    ),
    model,
  );

/** MiniMax expects reasoning_details as structured objects, not strings — never round-trip our split copy. */
const strip_minimax_reasoning_details = (msgs: message[], model: string): message[] => {
  if (!is_minimax_model(model)) return msgs;
  return msgs.map((msg) => {
    if (msg.role !== "assistant") return msg;
    const next = { ...msg };
    delete next.reasoning_details;
    return next;
  });
};

export const sanitize_tool_call_arguments = (msgs: message[]): message[] => {
  if (msgs.length === 0) return msgs;
  return msgs.map((msg) => {
    if (msg.role !== "assistant" || msg.tool_calls === undefined || msg.tool_calls.length === 0) return msg;
    return {
      ...msg,
      tool_calls: msg.tool_calls.map((tc) => ({ ...tc, function: { ...tc.function, arguments: valid_tool_arguments_json(tc.function.arguments) } })),
    };
  });
};

export const valid_tool_arguments_json = (raw: string): string => {
  raw = raw.trim();
  if (raw === "") return "{}";
  try {
    const obj = JSON.parse(raw) as unknown;
    if (obj === null || typeof obj !== "object" || Array.isArray(obj)) return "{}";
    return JSON.stringify(obj);
  } catch { return "{}"; }
};

const strip_images_if_needed = (msgs: message[], model: string): message[] => {
  if (supports_images(model)) return msgs;
  return msgs.map((msg) => {
    const blocks = content_blocks(msg);
    if (blocks === null || blocks.length === 0) return msg;
    const filtered: content_block[] = [];
    let image_count = 0;
    for (const b of blocks) {
      if (is_image_content_block(b)) image_count++;
      else filtered.push(b);
    }
    if (image_count === 0) return msg;
    const note = `[${image_count} image(s) omitted — this session does not support images.]`;
    if (filtered.length > 0 && filtered[0].type === "text") filtered[0].text = `${filtered[0].text ?? ""}\n${note}`;
    else filtered.unshift({ type: "text", text: note });
    const out = { ...msg };
    if (filtered.length === 0) out.content = new Uint8Array();
    else if (filtered.length === 1 && filtered[0].type === "text") out.content = string_content(filtered[0].text ?? "");
    else out.content = blocks_content(filtered);
    return out;
  });
};

const is_image_content_block = (b: content_block): boolean => b.type === "image_url" || b.type === "image";

export const repair_tool_messages = (msgs: message[]): message[] => {
  const out: message[] = [];
  for (let i = 0; i < msgs.length; i++) {
    const msg = { ...msgs[i] };
    // Orphan tool result with no preceding assistant tool_calls (e.g. the
    // issuing turn was dropped by trimming/compaction). Providers reject this
    // ("Messages with role 'tool' must be a response to a preceding message
    // with 'tool_calls'"), so drop it. Legitimate tool results are consumed in
    // the block right after their assistant tool_calls below.
    if (msg.role === "tool") continue;
    if (msg.role === "assistant" && msg.tool_calls !== undefined && msg.tool_calls.length > 0) {
      msg.tool_calls = msg.tool_calls.map((tc, j) => ({ ...tc, id: tc.id === "" ? `call_${i}_${j}` : tc.id }));
    }
    out.push(msg);
    if (msg.role !== "assistant" || msg.tool_calls === undefined || msg.tool_calls.length === 0) continue;
    const required: tool_call[] = [...msg.tool_calls];
    const valid_ids = new Set(required.map((tc) => tc.id));
    const answered = new Set<string>();
    i++;
    while (i < msgs.length && msgs[i].role === "tool") {
      const tm = { ...msgs[i] };
      let id = tm.tool_call_id;
      if (id === undefined || id === "") id = required[0].id;
      // Drop tool results whose id doesn't match this assistant's tool_calls
      // (mismatched orphan) or that we've already included (duplicate).
      if (!valid_ids.has(id) || answered.has(id)) {
        i++;
        continue;
      }
      tm.tool_call_id = id;
      out.push(tm);
      answered.add(id);
      i++;
    }
    i--;
    for (const tc of required) {
      if (answered.has(tc.id)) continue;
      out.push({ role: "tool", tool_call_id: tc.id, name: tc.function.name, content: string_content(tool_incomplete_msg) });
    }
  }
  return out;
};

export { normalize_messages };

